### PR TITLE
Log error if Okta and DB role claims unequal

### LINKED ISSUE #7598: Initiate rolling migration, phase 2
### Background
We would like to migrate user roles, org membership, and facility membership to be stored within the SimpleReport app, as opposed to managing it in Okta. This will increase stability of the app and improve code readability and extensibility, among other benefits. The goal of this ticket is to continue moving authorization data we access in Okta to our tables and also compare the data we've collected with our source of truth (Okta) to check for discrepancies.

### Change requested 
Change the code that accesses user role and facility/org membership from Okta to instead get this data from our tables if the feature flag is on.

If the feature flag is off, we continue to retrieve from Okta but we also update our data. Create entries in the ApiUserFacility and ApiUserRole tables or update existing entries.
When retrieving the data from Okta, compare to what we have in our tables and alert if they are not in sync.

### Acceptance criteria
- When the feature flag is off:
  - the behavior of the app re: authorization and showing role/facility info should reflect what's stored in Okta
  - any time the app performs an action that retrieves role or facility or org membership from Okta, the tables should also reflect this information
- When the feature flag is on:
  - the behavior of the app re: authorization and showing role/facility info should reflect what's stored in our tables
  - any time the app performs an action that retrieves role or facility or org membership, we do not update our tables
- (???) When the feature flag is off OR always until cleaned up
  - any time the app performs an action that retrieves role or facility or org membership, we compare what we get from Okta to what's stored in our tables. if the data doesn't match (inc is absent?), trigger an alert

### Dependencies
* #7596 
* #7597 

### Open questions
Do we want an alert if the data was absent or just if it was already in our table and doesn't match?
Question from earlier discussion: Do we want to do the comparison and alert when the flag is on?
Question from earlier discussion: For testing, any prod-specific corner cases we want to talk about that require special testing outside of testing in lowers?

### Notes
Test auth in lowers with non-superadmin (since superadmin still managed within Okta)

### Additional context
[Main design doc](https://docs.google.com/document/d/1YhRh9UIWj1kWeT35tj7T5pnUjW6Rn7qyIVkiQL3cDFU/edit?usp=sharing)
[Design doc - backend](https://docs.google.com/document/d/1NB6I9y8C2DdaPQSitMxJQ4rcWn63IMZd_c5eqU3RGyY/edit?usp=sharing)
[Design doc - data migration plan](https://docs.google.com/document/d/1UeHk7lBZI-dy__q_ibPxqkuaCUhofkByUTrIlmc3o30/edit?usp=sharing)
[Okta migration tickets plan](https://docs.google.com/document/d/1VnTLUFNvyo1coPHxcdkfimqW3wb9sgUzTdOrpgi-C3Y/edit?usp=sharing)
[Okta tech talk](https://docs.google.com/presentation/d/1LTrF3tmovKccN_ZUczXDjmhSUZkUsy1NNzYcrUzcuB8/edit?usp=sharing)
[Okta tech talk part 2](https://docs.google.com/presentation/d/1LhR17DEdGcpKtpf-BqWMyyWz_FHhgvE8PLcPtIPxRAI/edit?usp=sharing)

### PR DESCRIPTION
**Pull request by emyl3**
# BACKEND PULL REQUEST

## Related Issue
- FINALLY resolves #7598 😭 🙌 🎉 

## Changes Proposed
- Compare role claims in the DB and Okta and log an error if they are unequal for the following scenarios:
   - call `getUser` method
   - call `getUserByLoginEmail` method
   - when we set the `organizationRolesContext`

## Additional Information
- Two other issues were open related to finishing up issue #7598 
  - Adding a daily alert #8181 
  - Enabling the feature flag on certain dev envs #8180  

## Testing
- deployed on dev3 with the feature flag on
- Removed the following group in Okta for this user `cekok63012@nmaller.com` on `dev3`:
  - `SR-DEV3-TENANT:AK-Bobans-org-5145a894-3be7-45ea-a5a1-79895df0352c:ALL_FACILITIES`
  - searched the user in the "Manage users" support admin tool
  - [logged here](https://portal.azure.com/#view/AppInsightsExtension/DetailsV2Blade/ComponentId~/%7B%22SubscriptionId%22%3A%227d1e3999-6577-4cd5-b296-f518e5c8e677%22%2C%22ResourceGroup%22%3A%22prime-simple-report-dev%22%2C%22Name%22%3A%22prime-simple-report-dev3-insights%22%2C%22LinkedApplicationType%22%3A0%2C%22ResourceId%22%3A%22%252Fsubscriptions%252F7d1e3999-6577-4cd5-b296-f518e5c8e677%252FresourceGroups%252Fprime-simple-report-dev%252Fproviders%252FMicrosoft.Insights%252Fcomponents%252Fprime-simple-report-dev3-insights%22%2C%22ResourceType%22%3A%22microsoft.insights%252Fcomponents%22%2C%22IsAzureFirst%22%3Afalse%7D/DataModel~/%7B%22eventId%22%3A%225a28c81a-87ff-11ef-ac1d-000d3a52dd2e%22%2C%22timestamp%22%3A%222024-10-11T18%3A33%3A31.998Z%22%2C%22cacheId%22%3A%22f00a3d6a-af4c-4ca2-acf0-b57328bb6875%22%2C%22eventTable%22%3A%22traces%22%2C%22timeContext%22%3A%7B%22durationMs%22%3A86400000%2C%22endTime%22%3A%222024-10-11T18%3A34%3A01.398Z%22%7D%7D) in Azure
Screenshot 2024-10-11 at 13 35 15
 
 - log in as a non-site admin user on the envs above and navigate around the app

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java [hunk 1]
```diff
@@ -60,6 +60,8 @@ public List<OrganizationRoleClaims> findAllOrganizationRoles() {
       String username = currentAuth.getName();
       List<OrganizationRoleClaims> dbOrgRoleClaims =
           _dbOrgRoleClaimsService.getOrganizationRoleClaims(username);
+      _dbOrgRoleClaimsService.checkOrgRoleClaimsEquality(
```
Comment (emyl3) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java:63]: ⚠️ Anyone know how often we reset `organizationRolesContext` and have to fetch it again? 🤔 

Based on [digging around in Azure](https://portal.azure.com/#@cdc.onmicrosoft.com/resource/subscriptions/7d1e3999-6577-4cd5-b296-f518e5c8e677/resourceGroups/prime-simple-report-dev/providers/Microsoft.Insights/components/prime-simple-report-dev2-insights/searchV1), it seems like on initial log in and page load `fetchCurrentOrganizationRoles` (which is the only method that calls `findAllOrganizationRoles` is called a bunch of times and then not again for 20-30 mins or so... 

I'm asking because I'm wondering if we are doing the check too frequently if I set the check at this point. 🤔

File: backend/src/test/java/gov/cdc/usds/simplereport/service/ApiUserServiceTest.java [hunk 2]
```diff
@@ -159,7 +159,7 @@ void getUser_withAdminUser_withOktaMigrationDisabled_success() {
 
     UserInfo userInfo = _service.getUser(apiUser.getInternalId());
 
-    verify(_dbOrgRoleClaimsService, times(0)).getOrganizationRoleClaims((ApiUser) any());
+    verify(_dbOrgRoleClaimsService, times(1)).getOrganizationRoleClaims((ApiUser) any());
```
Comment (emyl3) [inline backend/src/test/java/gov/cdc/usds/simplereport/service/ApiUserServiceTest.java:162]: ⚠️ Had to increment by 1 in this test file because of how we are comparing the role claims -- we fetch from the DB regardless of whether the feature flag is on or off.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java (hunk 1 above)
Comment (DanielSass) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java:63]: Don't know off the top of my head.  And not sure if the better place to look would be spring docs or okta's 🤔

**Review by DanielSass — APPROVED** [review]
lgtm

**Review by bobbywells52 — APPROVED** [review]

File: backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java (hunk 1 above)
Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java:63]: does this line mean that the context is only valid for a single request? if so it seems like it would need to be refetched any time we are handling a new request that needs the current org roles
https://github.com/CDCgov/prime-simplereport/blob/f800b082b3ee8349e2747faef417b095f8613bf2/backend/src/main/java/gov/cdc/usds/simplereport/api/CurrentOrganizationRolesContextHolder.java#L11

Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java:63]: fwiw since we are condensing the logs into a single alert per day (?) it doesn't seem that harmful if we are logging frequently. though it may be a little bit of a pain to parse through the log lines later I don't see a super simple alternative 🤷

**Review by mehansen — APPROVED** [review]

File: backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java (hunk 1 above)
Comment (emyl3) [reply backend/src/main/java/gov/cdc/usds/simplereport/service/LoggedInAuthorizationService.java:63]: Thank you for digging into that @mehansen. Based on what you shared, I'm surprised I didn't see it log more frequently when testing. 🤷 After I merge, I'll keep on eye on prod and the logs...
